File: src/cosmic_pi_network/cpp/KnowledgeGraph.cpp
#include "KnowledgeGraph.h"
#include <networkx/networkx.h>

KnowledgeGraph::KnowledgeGraph() {
    this->graph = new nx::MultiDiGraph();
}

void KnowledgeGraph::add_entity(const char* entity) {
    graph->add_node(entity);
}

void KnowledgeGraph::add_relation(const char* entity1, const char* entity2, const char* relation) {
    graph->add_edge(entity1, entity2, relation);
}

void KnowledgeGraph::update(const char** new_entities) {
    for (int i = 0; new_entities[i]!= nullptr; i++) {
        add_entity(new_entities[i]);
    }
}

const char** KnowledgeGraph::get_entities() {
    int num_entities = graph->number_of_nodes();
    char** entities = new char*[num_entities];
    int i = 0;
    for (auto node : graph->nodes()) {
        entities[i++] = node.first.c_str();
    }
    return entities;
}

KnowledgeGraph::~KnowledgeGraph() {
    delete graph;
}

File: src/cosmic_pi_network/agi_framework/knowledge_graph.py
import networkx as nx

class KnowledgeGraph:
    def __init__(self):
        self.graph = nx.MultiDiGraph()

    def add_entity(self, entity):
        self.graph.add_node(entity)

    def add_relation(self, entity1, entity2, relation):
        self.graph.add_edge(entity1, entity2, relation=relation)

    def update(self, new_entities):
        for entity in new_entities:
            self.add_entity(entity)

    def get_entities(self):
        return list(self.graph.nodes)

    def __str__(self):
        return str(self.graph)

    def __repr__(self):
        return repr(self.graph)
